Source/WebCore/page/Navigation.cpp

```diff
@@ -97,37 +118,109 @@ Navigation::Result Navigation::reload(ReloadOptions&&, Ref<DeferredPromise>&& co
-    finished->resolve<IDLInterface<NavigationHistoryEntry>>(entry.get());
+    committed->resolve<IDLInterface<NavigationHistoryEntry>>(entry);
+    finished->resolve<IDLInterface<NavigationHistoryEntry>>(entry);
     return result;
 }
 
+std::optional<Ref<NavigationHistoryEntry>> Navigation::findEntryByKey(const String& key)
+{
+    auto entryIndex = m_entries.findIf([&key](const Ref<NavigationHistoryEntry> entry) {
+        return entry->key() == key;
+    });
+
+    if (entryIndex == notFound)
+        return std::nullopt;
+
+    return m_entries[entryIndex];
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-traverseto
+Navigation::Result Navigation::traverseTo(const String& key, Options&&, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
+{
+    auto current = currentEntry();
+    if (current && current->key() == key) {
+        auto globalObject = committed->globalObject();
+        Navigation::Result result = { DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(committed->promise())), DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(finished->promise())) };
+        committed->resolve<IDLInterface<NavigationHistoryEntry>>(*current);
+        finished->resolve<IDLInterface<NavigationHistoryEntry>>(*current);
+        return result;
+    }
+
+    auto entry = findEntryByKey(key);
+    if (!entry)
+        return createEarlyErrorResult(committed, finished, ExceptionCode::InvalidStateError, "Invalid key"_s);
+
+    return performTraversal(*entry, committed, finished);
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-back
+Navigation::Result Navigation::back(Options&&, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
+{
+    if (m_currentEntryIndex <= 0)
+        return createEarlyErrorResult(committed, finished, ExceptionCode::InvalidStateError, "Cannot go back"_s);
+
+    return performTraversal(m_entries[m_currentEntryIndex - 1], committed, finished);
+}
+
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-forward
 Navigation::Result Navigation::forward(Options&&, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
 {
-    // FIXME: keep track of promises to resolve later.
-    Ref entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
-    auto globalObject = committed->globalObject();
-    Navigation::Result result = { DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(committed->promise())), DOMPromise::create(*globalObject, *JSC::jsCast<JSC::JSPromise*>(finished->promise())) };
-    committed->resolve<IDLInterface<NavigationHistoryEntry>>(entry.get());
-    finished->resolve<IDLInterface<NavigationHistoryEntry>>(entry.get());
-    return result;
+    RELEASE_ASSERT(m_entries.size() <= INT_MAX);
+    if (m_currentEntryIndex < 0 || m_currentEntryIndex == static_cast<int>(m_entries.size()))
```

Comment:
I think the comparison may be wrong, for 1 entry we'd get 0 == 1 and not throw an exception, yet we still should not be allowed to go forward.
I had the same problem in https://github.com/WebKit/WebKit/pull/23578, you can make m_currentEntryIndex std::optional as a fix. For one thing the spec seems to have 1-based arrays, and then the above check would just work, but with 0-based we have to do size -1, thing about unsigned vs signed etc. Probably best to avoid such problems.